File: tmb_examples/TMBad/interpol.cpp
// Demonstrate 2D interpolation operator
#include <TMB.hpp>

template<class Type>
Type objective_function<Type>::operator() ()
{
  using tmbutils::interpol2D;
  using tmbutils::interpol2D_config;
  // Data for interpolation
  DATA_MATRIX(A);
  DATA_VECTOR(x_range);
  DATA_VECTOR(y_range);
  // Test evaluation for theese values
  PARAMETER_VECTOR(x);
  PARAMETER_VECTOR(y);  
  DATA_SCALAR(R);
  interpol2D_config<Type> cfg(R);
  interpol2D<Type> op(A, x_range, y_range, cfg);
  vector<Type> f = op(x, y);
  REPORT(f);
  return f.sum();
}

File: tmb_examples/TMBad/interpol.R
library(TMB)

## Compile and load the model
compile("interpol.cpp", framework="TMBad")
dyn.load(dynlib("interpol"))

## Smooth representation of 'volcano' data
data(volcano)
data <- list(A = volcano,
             x_range=c(0,1),
             y_range=c(0,1),
             R=2)
parameters <- list(x=0, y=0)

## Test different values of interpolation radius (R)
myf <- function(x, y, R) {
    data$R <- R
    parameters <- list(x=x, y=y)
    obj <- MakeADFun(data, parameters, DLL="interpol", silent=TRUE, type="double")
    obj$report(c(x,y))$f
}
layout(matrix(1:4,2))
x <- seq(.2,.4,length=100)
y <- seq(.2,.4,length=100)
for (R in 1:4) {
    z <- outer(x,y,myf, R=R)
    image(x,y,z)
    contour(x,y,z,add=TRUE)
    title(paste0("R=",R))
}

## Check higher order derivatives
obj <- MakeADFun(data, parameters, DLL="interpol", silent=TRUE)
library(numDeriv)
p <- c(.5, .5)
obj$fn(p)
obj$gr(p)
obj$gr(p) - grad(obj$fn, p)
obj$he(p) - jacobian(obj$gr, p)

## Optimize
opt <- optim(c(0.2,0.2), obj$fn, obj$gr, control=list(fnscale=-1))
